### ISSUE TITLE
Play with TOFU codebase

### ISSUE THREAD

**Issue by jack89roberts** [alan-turing-institute/arc-selective-forgetting#4]
See if we can replicate the whole TOFU pipeline, e.g. fine-tuning, unlearning, evaluation using their codebase (maybe with a small model rather than trying to replicate it on Phi/Llama).

**Comment by jack89roberts** [comment]
Writing some notes about the codebase here:

### General

- MIT license so we're free to do what we want with it
- Uses `hydra` for config management: https://hydra.cc/docs/1.3/intro/
- Uses `torchrun` for running jobs: https://pytorch.org/tutorials/beginner/ddp_series_fault_tolerance.html

### [Data modules](https://github.com/locuslab/tofu/blob/main/data_module.py)

- All the classes below use [`convert_raw_data_to_model_format` in `data_module.py`](https://github.com/locuslab/tofu/blob/8889542f281f7fca9ad23dbc11a4cb253ee2aa65/data_module.py#L8) to add model-specific question and answer tags to the data, e.g. `[INST] ` and `[/INST]` for Llama 2 (these are defined in [`config/model_config.yaml`](https://github.com/locuslab/tofu/blob/main/config/model_config.yaml)), and then tokenize the questions/answers (combined into one string).
- [TextDatasetQA](https://github.com/locuslab/tofu/blob/8889542f281f7fca9ad23dbc11a4cb253ee2aa65/data_module.py#L113) loads a requested dataset split and returns preprocessed, tokenized question/answers.
- [TextForgetDatasetQA](https://github.com/locuslab/tofu/blob/8889542f281f7fca9ad23dbc11a4cb253ee2aa65/data_module.py#L36) (used when doing unlearning) loads a requested `forget` split (e.g. `forget10`) and the corresponding retain split (e.g. `retain90`). It returns 2 concatenated samples each time when the data is accessed, either:
  - The requested index from the `forget` set and a random sample from the `retain` set
  - The requested index from the `idk` (I don't know) set and a random sample from the `retain` set
    - The `idk` sample is created by replacing the answer in the `forget` set with a random "I don't know"-style response from [`data/idontknow.jsonl`](https://github.com/locuslab/tofu/blob/main/data/idontknow.jsonl)
- [TextForgetDatasetDPOQA](https://github.com/locuslab/tofu/blob/8889542f281f7fca9ad23dbc11a4cb253ee2aa65/data_module.py#L76) returns 3 samples at a time (requested index from `forget` set, requested index from `idk` set, random sample from `retain` set)

### [Fine-tuning](https://github.com/locuslab/tofu/blob/main/finetune.py)

- Is a fairly standard HuggingFace causal language modelling trainer pipeline.
- With a [`CustomTrainer` class](https://github.com/locuslab/tofu/blob/8889542f281f7fca9ad23dbc11a4cb253ee2aa65/dataloader.py#L22) but not clear to what extent that's necessary - it has some minor modifications to what's returned by `compute_loss` and `prediction_step`.
- Option to use `LoRA` on the linear layers of a model (defaults in the readme do full fine-tuning).


### [Forgetting](https://github.com/locuslab/tofu/blob/main/forget.py)

- Uses `TextForgetDatasetQA` or `TextForgetDatasetDPOQA` (see above) and a fine-tuned model checkpoint as input.
- Has the option of LoRA on linear layers, same as fine-tuning.
- Uses [`CustomTrainerForgetting`](https://github.com/locuslab/tofu/blob/8889542f281f7fca9ad23dbc11a4cb253ee2aa65/dataloader.py#L45)
  - (Optionally) instantiated with a loss type, an "oracle model" (the oracle model is a frozen copy of the fine-tuned model, at least in the KL case), and an eval config.
    - The oracle model is used for KL divergence and DPO losses.
  - Overwrites [`compute_loss`](https://github.com/locuslab/tofu/blob/8889542f281f7fca9ad23dbc11a4cb253ee2aa65/dataloader.py#L92) to be able to do:
    - `grad_ascent`: Loss is the usual loss on the forget set * -1.
    - `grad_diff`: Loss is the `grad_ascent` loss + the loss on the retain set.
    - `KL`: `grad_ascent` loss + KL divergence between the oracle model and the current model on the retain set.
    - `idk`: Loss is the conventional loss on the `idk` set (forget set with answers replaced with "I don't know"-style strings)
    - `dpo`: Uses `forget` and `idk` losses with the current and oracle models.
  - Small modifications to `prediction_step`
  - `evaluate`: Sets up some model loading stuff with accelerate then computes and saves model utility,  retain result(?), forget quality.

### [Evaluation](https://github.com/locuslab/tofu/blob/main/evaluate_util.py)

- [`evaluate_util.py`]() seems to save some raw stats/metrics, and then [`aggregate_eval_stat.py`](https://github.com/locuslab/tofu/blob/main/aggregate_eval_stat.py) computes the aggregated metrics discussed in the paper.

**Comment by J-Dymond** [comment]
Can we close this issue now?
